### PR TITLE
[1LP][RFR] Cloud timeline adding events

### PR DESCRIPTION
**Pull request by t-fischer**
Refactoring and adding new event on the Cloud Timeline
=================
Rewriting the module the same way as test_timelines.py  and adding new events.

PRT failed because on some random events the source_instance field is missing ( a reported bug).
Some tests of the module failed because of some issues with providers/region. mostly overloaded, so events are not coming on time. On a run it might be good, on the next it might be delayed.
Test failed also on because somebody created a availability zone on Azure without a provider so but timeline button is disabled.

{{pytest: -v -q cfme/tests/cloud/test_cloud_timelines.py --use-template-cache --use-provider=complete}}

### CODE REVIEW

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 5ad17c4d3b274a72a1f365a96b96aaef0b088f2e:

`cfme/tests/cloud/test_cloud_timelines.py`:
- :red_circle: Line 8:1: **F401** *'cfme.exceptions.InstanceNotFound' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 93f679ab62ed97e200fc98681d8342d071da7213:

`cfme/tests/cloud/test_cloud_timelines.py`:
- :bangbang: Line 219:1: **E265** *block comment should start with '# '*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit da3a7e1da025036c831afc5b40413e0b83595eb4:

`cfme/cloud/instance/__init__.py`:
- :bangbang: Line 122:25: **E126** *continuation line over-indented for hanging indent*
- :bangbang: Line 128:25: **E126** *continuation line over-indented for hanging indent*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/tests/cloud/test_cloud_timelines.py [hunk 1]
```diff
@@ -1,147 +1,321 @@
...
+def test_cloud_timeline_create_event(new_instance, soft_assert, azone):
+    targets = (new_instance, new_instance.provider, azone)
+    event = 'create'
+    inst_event = InstEvent(new_instance, event)
+    logger.info('Will generate event %r on machine %r', event, new_instance.name)
+    wait_for(inst_event.emit, timeout='7m', message='Event {} did timeout'.format(event))
+    inst_event.catch_in_timelines(soft_assert, targets)
+
+
+@pytest.mark.uncollectif(lambda: current_version() < '5.9' and BZ(1542962),
```
Comment (quarckster) [inline cfme/tests/cloud/test_cloud_timelines.py]: **REQUIRED**
Please use `@pytest.mark.meta` for BZs

File: cfme/tests/cloud/test_cloud_timelines.py [hunk 2]
```diff
@@ -1,147 +1,317 @@
...
-    Metadata:
-        test_flag: timelines, provision
-    """
-    wait_for(count_events, [new_instance.provider, new_instance], timeout='5m', fail_condition=0,
-             message="events to appear")
+    profile = appliance.collections.policy_profiles.create(fauxfactory.gen_alpha(),
+                                                           policies=[policy])
+    request.addfinalizer(profile.delete)
 
+    new_instance.assign_policy_profiles(profile.description)
```
Comment (quarckster) [inline cfme/tests/cloud/test_cloud_timelines.py]: **REQUIRED**
Please convert it to yield fixture for readability sake.

File: cfme/tests/cloud/test_cloud_timelines.py [hunk 3]
```diff
@@ -1,147 +1,317 @@
...
     zone_id = new_instance.get_vm_via_rest().availability_zone_id
     zones = new_instance.appliance.rest_api.collections.availability_zones
     zone_name = next(zone.name for zone in zones if zone.id == zone_id)
-    azone = AvailabilityZone(name=zone_name, provider=new_instance.provider,
-                             appliance=new_instance.appliance)
-    wait_for(count_events, [azone, new_instance], timeout='5m', fail_condition=0,
-             message="events to appear")
+    inst_zone = AvailabilityZone(name=zone_name, provider=new_instance.provider,
+                                 appliance=new_instance.appliance)
+    yield inst_zone
```
Comment (quarckster) [inline cfme/tests/cloud/test_cloud_timelines.py]: **REQUIRED**
`yield` is not needed here.

File: cfme/tests/cloud/test_cloud_timelines.py [hunk 4]
```diff
@@ -1,147 +1,317 @@
...
+    logger.info('Will generate event %r on machine %r', event, new_instance.name)
+    wait_for(inst_event.emit, timeout='12m', message='Event {} did timeout'.format(event))
+    inst_event.catch_in_timelines(soft_assert, targets)
 
 
-@pytest.mark.manual
-def test_policy_events():
-    pass
+@pytest.mark.uncollectif(lambda provider: provider.one_of(EC2Provider) and
+                         current_version() < "5.9")
```
Comment (quarckster) [inline cfme/tests/cloud/test_cloud_timelines.py]: **REQUIRED**
```python
@pytest.mark.uncollectif(lambda provider, appliance: provider.one_of(EC2Provider) and appliance.version < "5.9")
```

File: cfme/tests/cloud/test_cloud_timelines.py [hunk 5]
```diff
@@ -1,147 +1,309 @@
...
-    wait_for(count_events, [new_instance.provider, new_instance], timeout='5m', fail_condition=0,
-             message="events to appear")
+    profile = appliance.collections.policy_profiles.create(fauxfactory.gen_alpha(),
+                                                           policies=[policy])
 
+    yield new_instance.assign_policy_profiles(profile.description)
+    request.addfinalizer(action.delete)
+    request.addfinalizer(policy.delete)
+    request.addfinalizer(policy.assign_events)
+    request.addfinalizer(profile.delete)
```
Comment (quarckster) [inline cfme/tests/cloud/test_cloud_timelines.py]: **REQUIRED**
Here as well.

File: cfme/tests/cloud/test_cloud_timelines.py [hunk 6]
```diff
@@ -1,147 +1,311 @@
...
+            timeline_filter.time_position.select_by_visible_text('centered')
+            timeline_filter.apply.click()
+            logger.info('Searching for event type: %r in timeline category: %r', self.event,
+                        self.tl_category)
+            return timelines_view.chart.get_events(self.tl_category)
+
+        events_list = _get_timeline_events(target, policy_events)
+        logger.debug('events_list: ', events_list)
+
+        if not len(events_list):
```
Comment (quarckster) [inline cfme/tests/cloud/test_cloud_timelines.py]: **OPTIONAL**
```python
if not events_list:
```

File: cfme/tests/cloud/test_cloud_timelines.py [hunk 7]
```diff
@@ -1,147 +1,311 @@
...
+    inst_event.catch_in_timelines(soft_assert, targets)
 
-    Metadata:
-        test_flag: timelines, provision
-    """
-    wait_for(count_events, [new_instance, new_instance], timeout='5m', fail_condition=0,
-             message="events to appear")
 
+def test_cloud_timeline_power_events(new_instance, soft_assert, azone):
+    events_list = ['stop', 'start']
```
Comment (quarckster) [inline cfme/tests/cloud/test_cloud_timelines.py]: **QUESTION**
Can this test be parametrized?

**Comment by quarckster** [comment]
@t-fischer please explain PRT failures.

File: cfme/tests/cloud/test_cloud_timelines.py (hunk 6 above)
Comment (t-fischer) [reply cfme/tests/cloud/test_cloud_timelines.py]: ok ll replace

File: cfme/tests/cloud/test_cloud_timelines.py (hunk 7 above)
Comment (t-fischer) [reply cfme/tests/cloud/test_cloud_timelines.py]: i ll divide the test in more part due to the pytest bug

File: cfme/tests/cloud/test_cloud_timelines.py [hunk 8]
```diff
@@ -1,147 +1,317 @@
...
+
+    def _power_off(self):
+        return self.inst.provider.mgmt.stop_vm(self.inst.name)
+
+    def _power_off_power_on(self):
+        return self._power_off() and self._power_on()
+
+    def _rename_vm(self):
+        logger.info('%r will be renamed', self.inst.name)
+        new_name = self.inst.provider.mgmt.set_name(self.inst.name, self.inst.name + '-renamed')
```
Comment (quarckster) [inline cfme/tests/cloud/test_cloud_timelines.py]: **REQUIRED**
Please use `format`

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: cfme/tests/cloud/test_cloud_timelines.py [hunk 9]
```diff
@@ -1,147 +1,320 @@
...
+from cfme.utils.wait import wait_for, TimedOutError
 
-
-cloud_prov_only = ProviderFilter(classes=[CloudProvider])
-excluded = ProviderFilter(classes=[GCEProvider, OpenStackProvider], inverted=True)
 pytestmark = [
     pytest.mark.tier(2),
-    pytest.mark.provider(gen_func=providers, filters=[excluded, cloud_prov_only], scope='module'),
+    # Only onr prov out of the 2 is taken, if not supplying --use-provider=complete
+    pytest.mark.provider([AzureProvider, EC2Provider], scope='module'),
```
Comment (psav) [inline cfme/tests/cloud/test_cloud_timelines.py]: **REQUIRED:** scope should be ```function``` for provider marker now

**Pull request merged**